ext/backtracie_native_extension/backtracie.c

```diff
@@ -16,105 +16,133 @@
-    if (stack_depth == 0 && !backtracie_is_thread_alive(thread)) return Qnil;
+// Get array of Backtracie::Locations for a given thread; if thread is nil,
+// returns for the current thread
+static VALUE collect_backtrace_locations(VALUE self, VALUE thread,
+                                         int ignored_stack_top_frames) {
+  if (!RTEST(thread)) {
+    thread = rb_thread_current();
   }
 
-  VALUE locations = rb_ary_new_capa(stack_depth - ignored_stack_top_frames);
+  // To maintain compatability with the Ruby thread backtrace behavior, if a
+  // thread is dead, then return nil.
+  if (!backtracie_is_thread_alive(thread)) {
+    return Qnil;
+  }
 
-  // MRI does not give us the path or line number for frames implemented using C code. The convention in
-  // Kernel#caller_locations is to instead use the path and line number of the last Ruby frame seen.
-  // Thus, we keep that frame here to able to replicate that behavior.
-  // (This is why we also iterate the frames array backwards below -- so that it's easier to keep the last_ruby_frame)
-  raw_location *last_ruby_location = 0;
+  int raw_frame_count = backtracie_frame_count_for_thread(thread);
 
-  for (int i = stack_depth - 1; i >= ignored_stack_top_frames; i--) {
-    VALUE location = Qnil;
+  // Allocate memory for the raw_locations, and keep track of it on the Ruby
+  // heap so it will be GC'd even if we raise.
+  // Zero the frame array so our mark function doesn't get confused too.
+  VALUE frame_wrapper = backtracie_frame_wrapper_new(raw_frame_count);
+  raw_location *raw_frames = backtracie_frame_wrapper_frames(frame_wrapper);
+  int *raw_frames_len = backtracie_frame_wrapper_len(frame_wrapper);
 
-    if (raw_locations[i].is_ruby_frame) {
-      last_ruby_location = &raw_locations[i];
-      location = ruby_frame_to_location(&raw_locations[i]);
-    } else {
-      location = cfunc_frame_to_location(&raw_locations[i], last_ruby_location);
+  for (int i = ignored_stack_top_frames; i < raw_frame_count; i++) {
+    bool valid_frame = backtracie_capture_frame_for_thread(
+        thread, i, &raw_frames[*raw_frames_len]);
+    if (valid_frame) {
+      (*raw_frames_len)++;
     }
+  }
 
-    rb_ary_store(locations, i - ignored_stack_top_frames, location);
+  VALUE rb_locations = rb_ary_new_capa(*raw_frames_len);
+  // Iterate _backwards_ through the frames, so we can keep track of the
+  // previous ruby frame for a C frame. This is required because C frames don't
+  // have filenames or line numbers; we must instead use the filename/lineno of
+  // the _caller_ of the function.
+  // note we seed prev_ruby_loc with the bottom frame so it's never null.
+  raw_location *prev_ruby_loc = &raw_frames[*raw_frames_len - 1];
+  for (int i = *raw_frames_len - 1; i >= 0; i--) {
+    if (raw_frames[i].is_ruby_frame) {
+      prev_ruby_loc = &raw_frames[i];
+    }
+    VALUE rb_loc = frame_to_location(&raw_frames[i], prev_ruby_loc);
```

Comment:
So I wrote a test for this - currently, if you call Backtracie from a brand new thread from rb_thread_create, it has a nil path for the top frame because of this logic. Calling Thread#backtrace_locations instead returns the name of the executable entrypoint (i.e. "rsepc-core-3.11.0/exe/rspec").

The null is potentially dangerous to applications not expecting it, but the entrypoint is a nonsensical thing to set it to too.

https://github.com/KJTsanaktsidis/backtracie/blob/73e141c5d624cc8333864634807e0f5390122c83/spec/unit/backtracie_spec.rb#L699

Should we...
* Keep it nil, since that's what it really is?
* Set it to the entrypoint, for compatability?
* Set it to some other not-null string, like "(cfunc)" or "(no path available)" or something that won't explode but isn't a lie?

Reply:
At least it doesn't crash :sweat_smile: 

> Should we...
> * Keep it nil, since that's what it really is?
> * Set it to the entrypoint, for compatability?
> * Set it to some other not-null string, like "(cfunc)" or "(no path available)" or something that won't explode but isn't a lie?

Excellent question. The reason thus far I've tried to always be able to match `backtrace_locations` is that it makes it much easier to test AND I have higher trust on that API being correct VS what `rb_profile_frames` sometimes gives us.

But to be honest here I'm leaning towards option #3: I think in this case it makes sense to break compatibility and add a placeholder. Perhaps `(in native code)`? but I'm OK with your other suggestions as well.

One thing we could do is represent this semantically in some way in the `raw_location`. Perhaps by having a `is_placeholder`? That way the caller could even discover that there's a placeholder and decide how it wants to encode it (e.g. wouldn't mask it as a regular frame called "(in native code").